### PR TITLE
Add new template to Truss

### PR DESCRIPTION
**Pull request by aspctu**

### CODE REVIEW

File: truss/templates/trtllm-whisper/model/model.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import requests
+import torch
+import logging
+import time
+
+
+class Model:
+    def __init__(self, **kwargs):
+        # Uncomment the following to get access
+        # to various parts of the Truss config.
```
Comment (pankajroark) [inline truss/templates/trtllm-whisper/model/model.py:15]: Not clear what to uncomment.

File: truss/templates/trtllm-whisper/model/model.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+        if audio_base64:
+            binary_data = base64.b64decode(audio_base64.encode("utf-8"))
+        elif url:
+            resp = requests.get(url)
+            binary_data = resp.content
+        return binary_data, request
+
+    async def predict(self, preprocessed_request) -> WhisperResult:
+        # Run model inference here
```
Comment (pankajroark) [inline truss/templates/trtllm-whisper/model/model.py:46]: Not necessary

File: truss/templates/trtllm-whisper/model/model.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        binary_data, request = preprocessed_request
+        waveform = self._model.preprocess_audio(binary_data)
+        return await self._model.transcribe(
```
Comment (pankajroark) [inline truss/templates/trtllm-whisper/model/model.py:49]: note to self: This needs to happen on threadpool, perhaps that's happening in transcribe.

**Review by pankajroark — COMMENTED** [review]
Still reviewing, will continue after meeting.

File: truss/config/trt_llm.py [hunk 4]
```diff
@@ -36,6 +37,7 @@ class CheckpointSource(Enum):
     HF: str = "HF"
     GCS: str = "GCS"
     LOCAL: str = "LOCAL"
+    REMOTE_URL: str = "REMOTE_URL"
```
Comment (pankajroark) [inline truss/config/trt_llm.py:40]: Good to add docs here for what this means.

File: truss/contexts/image_builder/serving_image_builder.py [hunk 5]
```diff
@@ -32,6 +32,9 @@
     TRTLLM_BASE_IMAGE,
     TRTLLM_PREDICT_CONCURRENCY,
     TRTLLM_PYTHON_EXECUTABLE,
+    AUDIO_MODEL_TRTLLM_REQUIREMENTS,
+    AUDIO_MODEL_TRTLLM_SYSTEM_PACKAGES,
+    AUDIO_MODEL_TRTLLM_TRUSS_DIR,
```
Comment (bdubayah) [inline truss/contexts/image_builder/serving_image_builder.py:37]: Missing in constants

File: truss/templates/trtllm-audio/packages/whisper_trt/batching.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+            tensors.append(tensors[-1])
+        res = torch.cat(tensors, dim=0).type(torch.float16)
+        return res
+
+    def concat_and_pad_prompts(self, prompts: list[list]) -> Tensor:
+        while len(prompts) < self.max_batch_size:
+            prompts.append(prompts[-1])
+        return Tensor(prompts)
+
+    def process_batch(self, batch: list[BatchWhisperItem]) -> list[float]:
```
Comment (pankajroark) [inline truss/templates/trtllm-audio/packages/whisper_trt/batching.py:32]: nit: list -> typing.List

File: truss/templates/trtllm-audio/packages/whisper_trt/batching.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        super().__init__(*args, **kwargs)
+        self.model: "WhisperModel" = model
+
+    def concat_and_pad_mels(self, tensors: list[Tensor]):
+        while len(tensors) < self.max_batch_size:
+            tensors.append(tensors[-1])
+        res = torch.cat(tensors, dim=0).type(torch.float16)
+        return res
+
+    def concat_and_pad_prompts(self, prompts: list[list]) -> Tensor:
```
Comment (pankajroark) [inline truss/templates/trtllm-audio/packages/whisper_trt/batching.py:27]: Doc comment indicating that modifications are done in place and using the past prompt_id.

File: truss/templates/trtllm-audio/packages/whisper_trt/batching.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        mel_batch = self.concat_and_pad_mels([item.mel for item in batch])
+
+        max_new_tokens = max(item.max_new_tokens for item in batch)
+        batch_result = self.model.process_batch(
+            mel_batch,
+            decoder_input_ids,
+            max_new_tokens=max_new_tokens,
+            num_beams=DEFAULT_NUM_BEAMS,
+        )
+        return batch_result[: len(batch)]
```
Comment (pankajroark) [inline truss/templates/trtllm-audio/packages/whisper_trt/batching.py:48]: May be a comment here explaining why the slice here is necessary/meaningful.

**Review by pankajroark — APPROVED** [review]
I don't have a deep understanding of the whisper model so couldn't do a good review of those portions. But I reviewed the parts about wiring this up into Truss and that lgtm.

File: truss/contexts/image_builder/serving_image_builder.py [hunk 9]
```diff
@@ -356,17 +365,23 @@ def copy_into_build_dir(from_path: Path, path_in_build_dir: str):
...
+                config.requirements.extend(
+                    BRITON_BASE_TRTLLM_REQUIREMENTS
+                    if USE_BRITON
+                    else BASE_TRTLLM_REQUIREMENTS
+                )
 
-            config.model_metadata["tags"] = [OPENAI_COMPATIBLE_TAG]
+                config.model_metadata["tags"] = [OPENAI_COMPATIBLE_TAG]
+            else:
+                config.requirements.extend(AUDIO_MODEL_TRTLLM_REQUIREMENTS)
```
Comment (bdubayah) [inline truss/contexts/image_builder/serving_image_builder.py:382]: What image ends up being used if no base image is specified / were you able to test the trtllm installation in it?

**Review by bdubayah — APPROVED** [review]
Looks good, one question I had: how much has the `whisper_trt` in packages diverged from `whisper_trt` in the FDE repo? I noticed some differences in variable names etc but seemed like core logic was unchanged

File: truss/contexts/image_builder/serving_image_builder.py (hunk 9 above)
Comment (aspctu) [reply truss/contexts/image_builder/serving_image_builder.py:382]: We don't use a base image here. The underlying base image is just the GPU base image that comes with Truss. Yes, the trtllm installation was tested.

**Pull request merged**